fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/CascadesPlanner.java

```diff
@@ -0,0 +1,406 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.temp;
+
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.record.RecordCoreArgumentException;
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.RecordMetaData;
+import com.apple.foundationdb.record.RecordStoreState;
+import com.apple.foundationdb.record.query.RecordQuery;
+import com.apple.foundationdb.record.query.plan.QueryPlanner;
+import com.apple.foundationdb.record.query.plan.plans.QueryPlan;
+import com.apple.foundationdb.record.query.plan.plans.RecordQueryPlan;
+import com.apple.foundationdb.record.query.plan.temp.expressions.NestedContextExpression;
+import com.apple.foundationdb.record.query.plan.temp.expressions.RelationalPlannerExpression;
+import com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.Iterator;
+
+/**
+ * A Cascades-style query planner that converts a {@link RecordQuery} to a {@link RecordQueryPlan}, possibly using
+ * secondary indexes defined in a {@link RecordMetaData} to execute the query efficiently.
+ *
+ * <p>
+ * <a href="https://15721.courses.cs.cmu.edu/spring2017/papers/15-optimizer2/graefe-ieee1995.pdf">Cascades</a> is a
+ * framework for a query optimization introduced by Graefe in 1995. In Cascades, all parsed queries, query plans, and
+ * intermediate state between the two are represented in a unified tree of {@link PlannerExpression}, which includes
+ * types such as {@link RecordQueryPlan} and {@link com.apple.foundationdb.record.query.expressions.QueryComponent}.
+ * This highly flexible data structure reifies essentially the entire state of the planner (i.e., partially planned
+ * elements, current optimization, goals, etc.) and allows individual planning steps to be modular and stateless by
+ * all state in the {@link PlannerExpression} tree.
+ * </p>
+ *
+ * <p>
+ * Like many optimization frameworks,Cascades is driven by {@link PlannerRule}s that describe a particular transformation
+ * and encapsulate the logic for detecting when it can be applied and how to app
+ * searches through its {@link PlannerRuleSet} to find a matching rule and then executes the rule, creating zero or more
+ * additional {@code PlannerExpression}s. In this implementation, a rule is defined by:
```

Comment:
I find the pronouns in this first sentence pretty confusing.

Reply:
Something still isn't quite right.

> Cascades is driven by {@link PlannerRule}s that describe a particular transformation and encapsulate the logic for detecting when it can be applied and how to apply them

Maybe "how to apply it?" (Otherwise, I think rules is the only possible referent for <i>them</i>.)

Reply:
Oh, I see. There's some terrible mixing of singular/plural here in general, which I think I've fixed now.